module/builder/collection/rate_limiter.go

```diff
@@ -0,0 +1,103 @@
+package collection
+
+import (
+	"math"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// rateLimiter implements payer-based rate limiting. See Config for details.
+type rateLimiter struct {
+
+	// maximum rate of transactions/payer/collection (from Config)
+	rate float64
+	// set of unlimited payer address (from Config)
+	unlimited map[flow.Address]struct{}
+	// height of the collection we are building
+	height uint64
+
+	// for each payer, height of latest collection in which a transaction for
+	// which they were payer was included
+	latestCollectionHeight map[flow.Address]uint64
+	// number of transactions included in the currently built block per payer
+	txIncludedCount map[flow.Address]uint
+}
+
+func newRateLimiter(conf Config, height uint64) *rateLimiter {
+	limiter := &rateLimiter{
+		rate:                   conf.MaxPayerTransactionRate,
+		unlimited:              conf.UnlimitedPayers,
+		height:                 height,
+		latestCollectionHeight: make(map[flow.Address]uint64),
+		txIncludedCount:        make(map[flow.Address]uint),
+	}
+	return limiter
+}
+
+// note the existence and height of a transaction in an ancestor collection.
+func (limiter *rateLimiter) addAncestor(height uint64, tx *flow.TransactionBody) {
+
+	// skip tracking payers if we aren't rate-limiting or are configured
+	// to allow multiple transactions per payer per collection
+	if limiter.rate >= 1 || limiter.rate <= 0 {
+		return
+	}
+
+	latest := limiter.latestCollectionHeight[tx.Payer]
+	if height >= latest {
```

Comment:
Is this comparison needed because a collection of a higher height could be built before (in time) another collection of lower height?

Reply:
No, this is to ensure that the value we eventually store in `latestCollectionHeight` is the largest of all the possible values we observe. 

We traverse the recent chain history and call this `addAncestor` function for every transaction. We want to determine for each payer account, what was the height of the most recent collection in which a transaction from that payer was included. 

If we don't have this check we might for example do something like:
* call `addAncestor` with `height=5, payer=P`
* call `addAncestor` with `height=4, payer=P`

In which case we would end up with `latestCollectionHeight[P] == 4`, when actually the most recent collection that P had a transaction in was at height 5.